File: src/main/java/com/github/kuro46/embedscript/script/processor/ScriptBuilder.java
package com.github.kuro46.embedscript.script.processor;

import com.github.kuro46.embedscript.script.Script;
import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.ImmutableListMultimap;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.ListMultimap;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

public class ScriptBuilder {
    private final UUID author;
    private Set<Script.MoveType> moveTypes = new HashSet<>();
    private Set<Script.ClickType> clickTypes = new HashSet<>();
    private Set<Script.PushType> pushTypes = new HashSet<>();
    private ListMultimap<String, String> script = ArrayListMultimap.create();

    private ScriptBuilder(UUID author) {
        this.author = author;
    }

    public static ScriptBuilder withAuthor(UUID author) {
        return new ScriptBuilder(author);
    }

    public UUID getAuthor() {
        return author;
    }

    public Set<Script.MoveType> getMoveTypes() {
        return moveTypes;
    }

    public Set<Script.ClickType> getClickTypes() {
        return clickTypes;
    }

    public Set<Script.PushType> getPushTypes() {
        return pushTypes;
    }

    public ListMultimap<String, String> getScript() {
        return script;
    }

    public Script build() {
        return new Script(author,
            ImmutableSet.copyOf(moveTypes),
            ImmutableSet.copyOf(clickTypes),
            ImmutableSet.copyOf(pushTypes),
            ImmutableListMultimap.copyOf(script));
    }
}

File: src/main/kotlin/com/github/kuro46/embedscript/script/processor/ScriptBuilder.kt
package com.github.kuro46.embedscript.script.processor

import com.github.kuro46.embedscript.script.Script
import com.google.common.collect.ArrayListMultimap
import com.google.common.collect.ImmutableListMultimap
import com.google.common.collect.ListMultimap
import java.util.HashSet
import java.util.UUID

/**
 * @author shirokuro
 */
class ScriptBuilder private constructor(val author: UUID) {
    val moveTypes: MutableSet<Script.MoveType> = HashSet()
    val clickTypes: MutableSet<Script.ClickType> = HashSet()
    val pushTypes: MutableSet<Script.PushType> = HashSet()
    val script: ListMultimap<String, String> = ArrayListMultimap.create()

    fun build(createdAt: Long): Script {
        return Script(author, createdAt,
                moveTypes.toHashSet(),
                clickTypes.toHashSet(),
                pushTypes.toHashSet(),
                ImmutableListMultimap.copyOf(script))
    }

    companion object {

        fun withAuthor(author: UUID): ScriptBuilder {
            return ScriptBuilder(author)
        }
    }
}
